cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java

```diff
@@ -95,68 +91,62 @@
    * @throws UnauthorizedException If the user does not have write permissions on the namespace.
    * @throws NamespaceNotFoundException If the specified namespace does not exist.
    * @throws AlreadyExistsException If the key already exists in the namespace. Updating is not supported.
    * @throws IOException If there was a problem storing the key to underlying provider.
    */
   @Override
-  public synchronized void put(SecureKeyId secureKeyId,
-                               SecureKeyCreateRequest secureKeyCreateRequest) throws Exception {
+  public final synchronized void putSecureData(String namespace, String name, String value, String description,
+                                               Map<String, String> properties) throws Exception {
     Principal principal = authenticationContext.getPrincipal();
-    NamespaceId namespaceId = new NamespaceId(secureKeyId.getNamespace());
+    NamespaceId namespaceId = new NamespaceId(namespace);
     authorizationEnforcer.enforce(namespaceId, principal, Action.WRITE);
 
-    String description = secureKeyCreateRequest.getDescription();
-    String value = secureKeyCreateRequest.getData();
     if (Strings.isNullOrEmpty(value)) {
       throw new BadRequestException("The data field should not be empty. This is the data that will be stored " +
                                       "securely.");
     }
 
-    secureStoreManager.putSecureData(secureKeyId.getNamespace(), secureKeyId.getName(), value, description,
-                           secureKeyCreateRequest.getProperties());
-    authorizer.grant(secureKeyId, principal, EnumSet.allOf(Action.class));
+    privilegesManager.grant(new SecureKeyId(namespace, name), principal, EnumSet.allOf(Action.class));
+    secureStoreManager.putSecureData(namespace, name, value, description, properties);
   }
 
   /**
    * Deletes the key if the user has ADMIN privileges to the key. Clears all the privileges associated with the key.
-   * @param secureKeyId Id of the key to be deleted.
    * @throws UnauthorizedException If the user does not have admin privileges required to delete the secure key.
    * @throws NamespaceNotFoundException If the specified namespace does not exist.
    * @throws NotFoundException If the key to be deleted is not found.
    * @throws IOException If there was a problem deleting it from the underlying provider.
    */
   @Override
-  public void delete(SecureKeyId secureKeyId) throws Exception {
+  public final void deleteSecureData(String namespace, String name) throws Exception {
     Principal principal = authenticationContext.getPrincipal();
+    SecureKeyId secureKeyId = new SecureKeyId(namespace, name);
     authorizationEnforcer.enforce(secureKeyId, principal, Action.ADMIN);
-    secureStoreManager.deleteSecureData(secureKeyId.getNamespace(), secureKeyId.getName());
-    authorizer.revoke(secureKeyId);
+    privilegesManager.revoke(secureKeyId);
```

Comment:
revoke should happen after deletion is successful, so there are no orphan secure keys (in case deletion fails) to which no one has access.